### PR TITLE
[W-11416105] Add transaction initiated from a source for prepare array and prepare struct

### PR DESCRIPTION
**Pull request by cristianchan-mulesoft**

### CODE REVIEW

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]
I've just make a couple of comments, after that, it is approved (IMHO)

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+begin
+    if z is not null and z.count > 1 then
+        z := TEST_STRUCT_COLLECTION(z(1));
+    end if;
+end;
+]]></db:sql>
+        </db:execute-ddl>
+    </munit:before-test>
+
+    <munit:test name="prepareArrayAndStructLocalTransactionSource" ignore="#[java!org::mule::extension::db::DbMunitUtils::isTestIgnored('oracle')]">
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:56]: Please could you add a test with a flow that triggers an error within the transaction?

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
+
+#### Compile  **Import add -D{vendor/profile}**
+```
+mvn clean install  -DruntimeProduct=MULE_EE -DruntimeVersion=4.3.0  -Dmunit.disable.lazy.initialization=false -Doracle
```
[the remark this answers is not in the document]
Comment (cristianchan-mulesoft) [reply README.md]: @acurci-at-mulesoft  done!

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml (hunk 1 above)
Comment (cristianchan-mulesoft) [reply src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:56]: Hey @acurci-at-mulesoft  these scenarios when the transaction fails are already cover in other tests.

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    </flow>
+
+    <flow name="dbListenerFlowExternalConnection">
+        <db:listener config-ref="${config}" table="TEST_LISTENER" transactionalAction="ALWAYS_BEGIN">
+            <scheduling-strategy>
+                <fixed-frequency  frequency="10"/>
+            </scheduling-strategy>
+        </db:listener>
+
+        <db:insert config-ref="oracleDbPoolingListenerExternalConfig">
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:150]: I think is accurate that the external operation also uses the prepareArray. Please, move it after the set variable.

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            </error-handler>
+        </try>
+        <try>
+            <db:execute-script config-ref="${config}">
+                <db:sql><![CDATA[DROP TABLE  TEST_LISTENER_EXTERNAL_CONNECTION]]></db:sql>
+            </db:execute-script>
+            <error-handler>
+                <on-error-continue enableNotifications="true" logException="false">
+                </on-error-continue>
+            </error-handler>
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:40]: Just so you know, these parts of code that you can use in the before and after sections to clean up the environment could be placed in a flow and then could be called from both sections with a flow-ref.

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        <db:execute-ddl config-ref="${config}">
+            <db:sql><![CDATA[CREATE TABLE TEST_LISTENER_EXTERNAL_CONNECTION (id number)]]></db:sql>
+        </db:execute-ddl>
+
+        <db:execute-ddl config-ref="${config}">
+            <db:sql>create or replace type TEST_STRUCT as object (field1 varchar2(3), field2 number(2), field3 varchar2(60))</db:sql>
+        </db:execute-ddl>
+        <db:execute-ddl config-ref="${config}">
+            <db:sql>create or replace type TEST_STRUCT_COLLECTION as table of TEST_STRUCT</db:sql>
+        </db:execute-ddl>
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:55]: It is a good practice to enclose the SQL sentence in a CDATA.

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        </munit:validation>
+    </munit:test>
+
+    <munit:after-suite name="dropStuff">
+        <try>
+            <db:execute-script config-ref="${config}">
+                <db:sql>
+                    drop procedure retain_first_struct;
+                    drop type TEST_STRUCT_COLLECTION;
+                    drop type TEST_STRUCT;
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:113]: add CDATA

File: src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                <fixed-frequency  frequency="10"/>
+            </scheduling-strategy>
+        </db:listener>
+
+        <set-variable
+                value="#[Db::prepareArray('TEST_STRUCT_COLLECTION', [ Db::prepareStruct('TEST_STRUCT', ['hi', 1, 'bye']), Db::prepareStruct('TEST_STRUCT', ['hi2', 2, 'bye2']) ])]"
+                variableName="list"/>
+
+        <db:stored-procedure config-ref="${config}" inOutParameters='#[{ "list": vars.list }]'>
+            <db:sql>call retain_first_struct(:list)</db:sql>
```
Comment (acurci-at-mulesoft) [inline src/test/munit/oracle/arrays-and-struct-source-oracle-test-case.xml:134]: add CDATA

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

**Review by acurci-at-mulesoft — APPROVED** [review]

**Pull request merged**